### ISSUE TITLE
issues with priovate key password in keystore

### ISSUE THREAD

**Issue by raffaelespazzoli** [redhat-cop/cert-utils-operator#112]
Looks like this may have inadvertently caused another issue. One of my developers is seeing an issue with the private key password.

Looking at the source code of cert-utils-operator which we use to generate keystore from tls secrets – I noticed that it is changed recently to fix frequent restarts of the pods every time the operator is restarted.. Below is the issue and the commit that fixes it.

``` 
https://github.com/redhat-cop/cert-utils-operator/issues/110
https://github.com/redhat-cop/cert-utils-operator/commit/c55ca62a3d9b67529476744c07eaa9097a648486#diff-266e243251ed42b2e1bcf54f2a041cdd69176644a9f4b4e2ec140edd32ea56e3L179
```
 
But the problem here is when the private key is being created and added to the keystore – it is added without any password – this line: https://github.com/redhat-cop/cert-utils-operator/blob/c55ca62a3d9b67529476744c07eaa9097a648486/controllers/secrettokeystore/secret_to_keystore_controller.go#L265
 
It needs to be changed to []byte(getPassword(secret)) instead of []byte{} – which means it is setting up empty password for the private key. And to test my hypothesis – I tried accessing the private key inside the keystore using an empty password and I was able to access it just fine.
 
So to sum it up – Keystore password is set to “jbosscert” and key password is set to empty string when the JKS files are generated.

[jdziedzic](https://github.com/jdziedzic) I am not sure I'm following. We were not setting the key password before the change.
Also is the expectation that we put the password for the Key? Why is the key-level password needed? 
If so is it going to be the same password as the keystore? I don't udnertsand how java apps work in this space, but we have only one field to specify a password, not two.
Also just to understand by your observation the way the code works now is that it generates an empty password for the key and `jbosscert` password for the keystore regardless of the password you set?

**Comment by ahnarayanan** [comment]
@raffaelespazzoli the previous behavior of the operator was that the private key generated from the secret and the keystore file were both using the same password which was set on the annotation `cert-utils-operator.redhat-cop.io/java-keystore-password` or `changeme`. 

Current behavior is that the keystore uses the right password we set but private key is using an empty string as the password. 

>  Also is the expectation that we put the password for the Key? Why is the key-level password needed?

Yes. setting password for the private key entries gives an extra layer of security to decrypt the private key after unlocking the keystore with the store password. In general key password inside the keystore is set to be same as that of the keystore password too, which is the default behavior that keytool utility follows as well. 

> but we have only one field to specify a password, not two.

I think that is ok. If we just make sure the same password is set in both cases, it will do, just like before. 

> We were not setting the key password before the change.

It also appears like the upstream keystore-go library [changed](https://github.com/pavel-v-chernykh/keystore-go/commit/adfaf7db24f828f8ed6734645925029775df45a1#diff-841beb1c3d91e0f9d97cd4e608c4b5c9357d2440709fe56c3832fc73614b76e3L8) how keystores are generated which may have changed the behavior as well. So at that time, the module used the keystore password as the key password but it has since been changed to support setting password for the private key which i think, have caused the issue we are seeing now. 

hope this helps.

**Comment by jdziedzic** [comment]
Any chance we could get a release cut for this now that its merged, so we can upgrade the operator in our OpenShift clusters?
